### PR TITLE
Course Refresh with TMC langs Rust

### LINKED ISSUE #467: If course is unrefreshable refresh process goes into infinite loop
This happens at least with the initial refresh.

### PR DESCRIPTION
**Pull request by sebazai**
Resolves #467 

- [x] Implement Refresh in Background daemon
- [x] ActionCable to report realtime progress to frontend
- [x] Make progress bar / progress updater of Rust background
- [x] Update exercise/course data in database
- [x] Generate report of added points and exericses etc. after progress
- [x] Improve Refresh report and progress report looks
- [x] Fix specs
  - [x] v8
    - [x] Requires tmc-langs to support individual exercises in repo root and zip them as name.zip and folder structure -> (name/src/SimpleStuff.java, name/test/SimpleStuffTest.java, etc.)
  - [x] models
  - [x] controllers
  - [x] features
  - [x] lib
  - [x] integration
- [x] config/environment/production.rb cable setup 
- [x] Remove phase 3 in course template creation
- [x] Refresh report/progress in course template site

### CODE REVIEW

File: app/background_tasks/refresh_course_task.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        courses.each do |course|
+          Rails.logger.info("Refreshing course #{course.name}")
+          RustLangsCliExecutor.refresh(course, task.id, task.no_background_operations, task.no_directory_changes)
+          finished_successfully = true
+        end
+      rescue => e
+        Rails.logger.error("Task failed: #{e}")
+        Rails.logger.error(e.backtrace.join("\n"))
+      end
+      task.destroy! if finished_successfully
```
Comment (sebazai) [inline app/background_tasks/refresh_course_task.rb:24]: Destroys entry created in database and all the phases.

Comment (nygrenh) [reply app/background_tasks/refresh_course_task.rb:24]: Yeah maybe someting like `.update(completed: true)` would be better here.

Comment (sebazai) [reply app/background_tasks/refresh_course_task.rb:24]: Yea, I will use the status from the CourseRefresh, i.e. Course.where(status: :not_started).each do |t|...

File: config/environments/production.rb [hunk 2]
```diff
@@ -18,6 +18,10 @@
   config.consider_all_requests_local       = false
   config.action_controller.perform_caching = true
 
+  config.action_cable.allowed_request_origins['https://tmc.mooc.fi']
```
Comment (sebazai) [inline config/environments/production.rb:21]: Could server url be fetched from some constant?

File: app/background_tasks/refresh_course_task.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        @refresh = CourseRefreshDatabaseUpdater.new.refresh_course(course, rust_output)
+      end
+
+      broadcast_to_channel(channel_id, 'Generating refresh report', 0.98, '-')
+      CourseTemplateRefreshReport.create(course_template_refresh_id: task.id, refresh_errors: @refresh.errors, refresh_warnings: @refresh.warnings, refresh_notices: @refresh.notices, refresh_timings: @refresh.timings)
+
+      broadcast_to_channel(channel_id, 'Cleaning up cache', 0.99, '-')
+      old_cache_path = task.course_template.cache_path
+      task.course_template.increment_cached_version
+      task.course_template.save!
```
Comment (sebazai) [inline app/background_tasks/refresh_course_task.rb:35]: check that task.course_template matches rust_output['output-data'][new-cache-path]

File: app/background_tasks/refresh_course_task.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      task.course_template.reload
+      courses.each(&:save!)
+      FileUtils.rm_rf(old_cache_path)
```
Comment (sebazai) [inline app/background_tasks/refresh_course_task.rb:38]: Remove in background task with wait_delay 2-3 mins?

File: app/views/courses/show.html.erb [hunk 5]
```diff
@@ -191,7 +199,13 @@
 
           <% # General info %>
           <% if @course.refreshed_at %>
-            <li>Last refreshed at: <%= @course.refreshed_at.strftime("%d.%m.%Y %H:%M:%S") %></li>
+            <li>
+              Last refreshed at <%= @course.refreshed_at.strftime("%d.%m.%Y %H:%M:%S") %> 
+              <% if @course.course_template.course_template_refreshes.any? %>
+                by <%= @course.course_template.course_template_refreshes.last.user.email %>
```
Comment (sebazai) [inline app/views/courses/show.html.erb:205]: Should this be visible on course page or only for admins in database?

Comment (sebazai) [reply app/views/courses/show.html.erb:205]: @nygrenh Could you comment plz.

**Pull request merged**